### PR TITLE
#1449 - Integration of the PRACC / HPRM Manager

### PR DESCRIPTION
**Pull request by kvanhijf**

### CODE REVIEW

File: ovs/lib/storagerouter.py [hunk 1]
```diff
@@ -1264,6 +1265,128 @@ def get_logfiles(local_storagerouter_guid):
...
 
     @staticmethod
+    @ovs_task(name='ovs.storagerouter.create_hprm_config_files')
+    def create_hprm_config_files(storagerouter_guid, local_storagerouter_guid, parameters):
+        """
+        Create the required configuration files to be able to make use of HPRM (aka PRACC)
+        These configuration will be zipped and made available for download
+        :param storagerouter_guid: The guid of the StorageRouter for which a HPRM manager needs to be deployed
+        :type storagerouter_guid: str
+        :param local_storagerouter_guid: The guid of the StorageRouter this call is executed on
```
Comment (khenderick) [inline ovs/lib/storagerouter.py]: It's not about this call, but where the API was executed

File: ovs/lib/storagerouter.py [hunk 2]
```diff
@@ -1264,6 +1265,128 @@ def get_logfiles(local_storagerouter_guid):
...
+
+        if config_path is None:
+            raise ValueError('vPool {0} has not been extended to StorageRouter {1}'.format(vpool.name, storagerouter.name))
+
+        alba_backend_name = None
+        fragment_cache_on_read = parameters['fragment_cache_on_read']
+        fragment_cache_on_write = parameters['fragment_cache_on_write']
+        if fragment_cache_on_read is True or fragment_cache_on_write is True:
+            if 'path' in parameters:  # Local caching
+                required_params.update({'path': (str, re.compile('^/$|/.*[^.]$')),
```
Comment (khenderick) [inline ovs/lib/storagerouter.py]: That doesn't make much sense to me, are you sure this works? I'd expect something like `^/.*[^.]?$`

File: ovs/lib/storagerouter.py [hunk 3]
```diff
@@ -1264,6 +1265,128 @@ def get_logfiles(local_storagerouter_guid):
...
+                required_params.update({'backend_info': (dict, {'preset': (str, Toolbox.regex_preset),
+                                                                'alba_backend_name': (str, Toolbox.regex_backend)}),
+                                        'connection_info': (dict, {'host': (str, Toolbox.regex_ip, False),
+                                                                   'port': (int, {'min': 1, 'max': 65535}, False),
+                                                                   'client_id': (str, Toolbox.regex_guid, False),
+                                                                   'client_secret': (str, None, False)})})
+                Toolbox.verify_required_params(actual_params=parameters,
+                                               required_params=required_params)
+                alba_backend_name = parameters['backend_info']['alba_backend_name']
+                if Configuration.dir_exists(key='/ovs/arakoon/{0}-abm'.format(alba_backend_name)) is False:
```
Comment (khenderick) [inline ovs/lib/storagerouter.py]: If it's an external cluster, you'll have to use the API calls like we use in the `add_vpool` logic.

File: ovs/lib/storagerouter.py [hunk 4]
```diff
@@ -1264,6 +1265,128 @@ def get_logfiles(local_storagerouter_guid):
...
+                                 'transport': 'tcp',
+                                 'log_level': 'info',
+                                 'fragment_cache': fragment_cache_info,
+                                 'read_preference': proxy_cfg['read_preference'],
+                                 'albamgr_cfg_url': '/etc/hprm/{0}/arakoon.ini'.format(vpool.name),
+                                 'manifest_cache_size': proxy_cfg['manifest_cache_size']})
+        file_contents_map = {'/opt/OpenvStorage/config/{0}/config.json'.format(vpool.name): config,
+                             '/opt/OpenvStorage/config/{0}/arakoon.ini'.format(vpool.name): Configuration.get(key=config_path.format('abm'), raw=True)}
+        if fragment_cache_info[0] == 'alba':
+            file_contents_map['/opt/OpenvStorage/config/{0}/accelerated_arakoon.ini'.format(vpool.name)] = Configuration.get(key='/ovs/arakoon/{0}-abm/config'.format(alba_backend_name), raw=True)
```
Comment (khenderick) [inline ovs/lib/storagerouter.py]: This should then be the actual config as loaded from the API.

File: webapps/api/backend/views/storagerouters.py [hunk 5]
```diff
@@ -153,6 +153,44 @@ def get_support_info(self, storagerouter):
...
+    @load(StorageRouter)
+    def get_proxy_config(self, storagerouter, vpool_guid):
+        """
+        Gets the ALBA proxy for a given Storage Router and vPool
+        :param storagerouter: StorageRouter on which the ALBA proxy is configured
+        :type storagerouter: StorageRouter
+        :param vpool_guid: Guid of the vPool for which the proxy is configured
+        :type vpool_guid: str
+        """
+        return StorageRouterController.get_proxy_config.s(vpool_guid=vpool_guid,
```
Comment (khenderick) [inline webapps/api/backend/views/storagerouters.py]: Isn't `.delay(...)` easier?

File: webapps/frontend/app/viewmodels/wizards/createhprmconfigs/data.js [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+// in the LICENSE.txt file of the Open vStorage OSE distribution.
+//
+// Open vStorage is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY of any kind.
+/*global define */
+define(['jquery', 'knockout'], function($, ko){
+    "use strict";
+    var pathRegex, hostRegex, singleton, parsePresets;
+
+    pathRegex = /^\/$|\/.*[^.]$/;
```
Comment (khenderick) [inline webapps/frontend/app/viewmodels/wizards/createhprmconfigs/data.js]: Same remark as about the regex above.

File: webapps/frontend/app/viewmodels/wizards/createhprmconfigs/gather.js [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+                        self.data.albaPreset(undefined);
+                        self.loadingBackends(false);
+                        self.loadingBackendsFailed(true);
+                        albaDeferred.reject();
+                    });
+            }).promise();
+        };
+        self.reloadBackends = function() {
+            self.reloadingBackends(true);
+            $.when.apply($, [self.loadBackends()])
```
Comment (khenderick) [inline webapps/frontend/app/viewmodels/wizards/createhprmconfigs/gather.js]: This is a bit strange.

**Review by khenderick — CHANGES_REQUESTED** [review]

**Review by khenderick — DISMISSED** [review]

**Comment by khenderick** [comment]
Discarded review, since in the meantime develop has evolved quite a bit, so there's some extra re-integration work.

File: ovs/lib/storagerouter.py [hunk 8]
```diff
@@ -1396,6 +1396,156 @@ def get_logfiles(local_storagerouter_guid):
         client.run(['chmod', '666', '{0}/{1}'.format(webpath, logfilename)])
         return logfilename
 
+    # noinspection PyTypeChecker
+    @staticmethod
+    @ovs_task(name='ovs.storagerouter.create_hprm_config_files')
+    def create_hprm_config_files(storagerouter_guid, local_storagerouter_guid, parameters):
+        """
+        Create the required configuration files to be able to make use of HPRM (aka PRACC)
+        These configuration will be zipped and made available for download
```
Comment (JeffreyDevloo) [inline ovs/lib/storagerouter.py]: This* configuration

File: ovs/lib/storagerouter.py [hunk 9]
```diff
@@ -1396,6 +1396,156 @@ def get_logfiles(local_storagerouter_guid):
...
+                    raise ValueError('No ALBA proxies configured for vPool {0} on StorageRouter {1}'.format(vpool.name, storagerouter.name))
+                config_path = '/ovs/vpools/{0}/proxies/{1}/config/{{0}}'.format(vpool.guid, sd.alba_proxies[0].guid)
+
+        if config_path is None:
+            raise ValueError('vPool {0} has not been extended to StorageRouter {1}'.format(vpool.name, storagerouter.name))
+        proxy_cfg = Configuration.get(key=config_path.format('main'))
+
+        cache_info = {}
+        arakoons = {}
+        cache_types = ['fragment', 'block']
```
Comment (JeffreyDevloo) [inline ovs/lib/storagerouter.py]: Could we define these somewhere in an enum? I figure it has been used on some more places.

File: ovs/lib/storagerouter.py [hunk 10]
```diff
@@ -1396,6 +1396,156 @@ def get_logfiles(local_storagerouter_guid):
...
+                    path = parameters[ctype].get('path')
+                    if path is not None:
+                        path = path.strip()
+                        if not path or path.endswith('/.') or '..' in path or '/./' in path:
+                            raise ValueError('Invalid path specified')
+                        required_params[ctype][1].update({'path': (str, None),
+                                                          'size': (int, {'min': 1, 'max': 10 * 1024})})
+                        Toolbox.verify_required_params(actual_params=parameters,
+                                                       required_params=required_params)
+                        while '//' in path:
```
Comment (JeffreyDevloo) [inline ovs/lib/storagerouter.py]: No need to iterate here, by default replace will replace all occurences

File: ovs/lib/storagerouter.py [hunk 11]
```diff
@@ -1396,6 +1396,156 @@ def get_logfiles(local_storagerouter_guid):
...
+                                               required_params=required_params)
+                read = parameters[ctype]['read']
+                write = parameters[ctype]['write']
+                if read is True or write is True:
+                    path = parameters[ctype].get('path')
+                    if path is not None:
+                        path = path.strip()
+                        if not path or path.endswith('/.') or '..' in path or '/./' in path:
+                            raise ValueError('Invalid path specified')
+                        required_params[ctype][1].update({'path': (str, None),
```
Comment (JeffreyDevloo) [inline ovs/lib/storagerouter.py]: Define an argument for referring to this dict to up the readability.

File: ovs/lib/storagerouter.py [hunk 12]
```diff
@@ -1396,6 +1396,156 @@ def get_logfiles(local_storagerouter_guid):
...
+            raise ValueError('At least one cache type should be passed: {0}'.format(', '.join(cache_types)))
+        for ctype in cache_types:
+            if ctype in parameters:
+                required_params.update({ctype: (dict, {'read': (bool, None),
+                                                       'write': (bool, None)})})
+                Toolbox.verify_required_params(actual_params=parameters,
+                                               required_params=required_params)
+                read = parameters[ctype]['read']
+                write = parameters[ctype]['write']
+                if read is True or write is True:
```
Comment (JeffreyDevloo) [inline ovs/lib/storagerouter.py]: if read is False and write is False:
 cache_info[ctype] = ['none']
 continue
INSERT THE CURRENT IF CODE HERE

File: ovs/lib/storagerouter.py [hunk 13]
```diff
@@ -1396,6 +1396,156 @@ def get_logfiles(local_storagerouter_guid):
...
+        file_contents_map.update({'/opt/OpenvStorage/config/{0}/config.json'.format(identifier): json.dumps(config, indent=4),
+                                  '/opt/OpenvStorage/config/{0}/arakoon.ini'.format(identifier): Configuration.get(key=config_path.format('abm'), raw=True)})
+
+        local_client = SSHClient(endpoint=local_storagerouter)
+        local_client.dir_create(directories='/opt/OpenvStorage/config/{0}'.format(identifier))
+        local_client.dir_create(directories='/opt/OpenvStorage/webapps/frontend/downloads')
+        for file_name, contents in file_contents_map.iteritems():
+            local_client.file_write(contents=contents,
+                                    filename=file_name)
+        local_client.run(command='tar --transform "s#^config/{0}#{0}#" -czf /opt/OpenvStorage/webapps/frontend/downloads/{1} config/{0}'.format(identifier, tgz_name), allow_insecure=True)
```
Comment (JeffreyDevloo) [inline ovs/lib/storagerouter.py]: Insecure necessary here?

**Review by JeffreyDevloo — CHANGES_REQUESTED** [review]

File: ovs/lib/storagerouter.py (hunk 10 above)
Comment (khenderick) [reply ovs/lib/storagerouter.py]: This small loop will also replace `/////` for example.

**Review by JeffreyDevloo — APPROVED** [review]

**Pull request merged**